### ISSUE TITLE
Keyboard CSS Selectors in Cameroon Qwerty.

### ISSUE THREAD

**Issue by MattGyverLee** [keymanapp/keyboards#1388]
Hi folks, 
The last time I was working on my keyboard, Marc recommended that I add ".kmw-keyboard-sil_cameroon_qwerty" to my CSS selectors that were "too powerful" with just the platform and key. My understanding was that they could "leak" to other keyboards. 

I just realized that changing the CSS to the recommended values back then (https://github.com/keymanapp/keyboards/commit/240399c3ed3b947a221612e62e81f71308d0f35a#diff-cba02f4140cde896c6de1a2b94c339c7c1dde11353018971aa189f8a3798087e) has broken the css in anything but "Test Keyboard on web" for the keyboard that was changed (Cameroon Qwerty). Cameroon Azerty's styling is working perfectly without the keyboard specification. 

In short, the following CSS selector works on Android/iOS:
`.android #shift-T_0330 { background: #598059; background: linear-gradient(#ffffff, #598059); }`
and the following only works on the web preview, but fails to work on "real" Android devices:
`.android .kmw-keyboard-sil_cameroon_qwerty #shift-T_0330 { background: #598059; background: linear-gradient(#ffffff, #598059); }`

You can compare the current Cameroon Azerty (styled) and Cameroon Qwerty (un-styled). Is there a proper keyboard selector for Android and iOS keyboards that I should be using that works in the KM app and KM Engine? Please advise.

**Comment by DavidLRowe** [comment]
sounds like a question for @mcdurdin or @jahorton ?

**Comment by mcdurdin** [comment]
I inspected on my Android device and saw the following class for the 'khmer_angkor' keyboard: `kmw-keyboard-khmer_angkor::khmer_angkor`. I don't know why we're getting that class name and it smells like a bug @jahorton?

That was with 14.0.184-alpha.

**Comment by jahorton** [comment]
> I inspected on my Android device and saw the following class for the 'khmer_angkor' keyboard: `kmw-keyboard-khmer_angkor::khmer_angkor`. I don't know why we're getting that class name and it smells like a bug @jahorton?
> 
> That was with 14.0.184-alpha.

No, that sounds right - that's how keyboards are 'namespaced' in the embedded context, at least on Android.  To be clear, that's `kmw-keyboard-<package id>::<keyboard id>`, a decision @darcywong00 made when we first started utilizing the KMP format as part of the Android implementation.  I'd have to double-check whether or not that's mirrored within the iOS engine.

**Comment by jahorton** [comment]
For reference, see https://github.com/keymanapp/keyman/blob/f16872a99896a2fc208c68ac92a827820ab07e3f/web/source/kmwembedded.ts#L51-L56 - the underlying cause of this, at least, has been in place since 10.0.

**Comment by jahorton** [comment]
I've registered a corresponding issue on the main repo.

**Comment by darcywong00** [comment]
> To be clear, that's kmw-keyboard-<package id>::<keyboard id>, resulting from a keyboard's full ID being <package id>::<keyboard id>, a decision @darcywong00 made when we first started utilizing the KMP format as part of the Android implementation

I believe that was going by the KMP installation [spec](
https://docs.google.com/document/d/1sj7W6pCiN-_iRss5iRdib1aHaSTmYoLIueQSKJeNy8Q/edit#)

> Add keyboard to language in manager's keyboard list, remembering namespacing: packageid.keyboardid for packaged keyboards, and just plain old keyboardid for legacy keyboards.

I think legacy keyboards that didn't have a package ID would be treated as cloud::keyboardid

**Comment by MattGyverLee** [comment]
The Package ID and keyboard with the KMP make sense (and I'm really liking using the KMP)...it's just new to me...and apparently @mcdurdin as well. 

Good discussion! Ok, so iOS expects `kmw-keyboard-khmer_angkor` and current android versions expect `kmw-keyboard-khmer_angkor::khmer_angkor`. I've reverted to not using the package name for now locally, 

.ios #shift-T_0330
.android #shift-T_0330

I need to push a new update to the repo (since the qwerty style is broken), so what do you suggest I put in my keyboard in the repo to make it current and forward compatible? Should I put in both for now in case you decide to change it?

.ios kmw-keyboard-sil_cameroon_qwerty #shift-T_0330
.android kmw-keyboard-sil_cameroon_qwerty::sil_cameroon_qwerty #shift-T_0330
.android kmw-keyboard-sil_cameroon_qwerty #shift-T_0330

Have you seen a too-powerful css style "leak" into another keyboard or is this a theoretical problem? The same key-names are styled in both keyboards, so even when it is styling powerfully, I wouldn't see a leak.,
Am I the only one styling keys this way?

On another note, I'm familiar with inspection in browsers, but how do you inspect in android? That could be useful.

**Comment by darcywong00** [comment]
> On another note, I'm familiar with inspection in browsers and inspection of my app, but how do you inspect the keyboard itself in android? That could be useful.

When you have your keyboard running (for example emulator in Android Studio), then open Chrome on your host PC
chrome://inspect/#devices

When the remote device connects, you'll see a Keyboard WebView that you can inspect.

**Comment by mcdurdin** [comment]
> `.ios kmw-keyboard-sil_cameroon_qwerty #shift-T_0330`
> `.android kmw-keyboard-sil_cameroon_qwerty::sil_cameroon_qwerty #shift-T_0330`
> `.android kmw-keyboard-sil_cameroon_qwerty #shift-T_0330`

If the iOS and Android styles are different, I'd use:

```
.ios kmw-keyboard-sil_cameroon_qwerty #shift-T_0330 {
}

.android kmw-keyboard-sil_cameroon_qwerty #shift-T_0330,
.android kmw-keyboard-sil_cameroon_qwerty::sil_cameroon_qwerty #shift-T_0330, /* support Keyman 10.0-13.0 bug #3910 */
.android kmw-keyboard-cloud::sil_cameroon_qwerty #shift-T_0330 /* support Keyman 13.0 bug #3910 */
{
  // css
}
```

If they are the same, then:

```
kmw-keyboard-sil_cameroon_qwerty #shift-T_0330,
kmw-keyboard-sil_cameroon_qwerty::sil_cameroon_qwerty #shift-T_0330, /* support Keyman 10.0-13.0 bug #3910 */
kmw-keyboard-cloud::sil_cameroon_qwerty #shift-T_0330 /* support Keyman 13.0 bug #3910 */
{
  // css
}
```

Yes, you are using CSS more extensively than the vast majority of developers. CSS tweaks are an area subject to change as we have never solidified the spec around custom CSS for keyboards. The flux may be more significant in upcoming releases as we look at possibilities of restricting the CSS or providing custom controls within Developer. There is also scope for third-party 'skinning' which may come into play as well.

> When you have your keyboard running (for example emulator in Android Studio), then open Chrome on your host PC
> chrome://inspect/#devices

Note that if you are doing this with your physical device, it will need to have debugging enabled and for some devices, you need to install gdb and specific usb drivers as well on your host. YMMV. Follow the help dots at the Chrome Inspector for more info.

**Comment by jahorton** [comment]
> On another note, I'm familiar with inspection in browsers and inspection of my app, but how do you inspect the keyboard itself in android? That could be useful.

If you have the Keyman app or its system keyboard launched on your Android device and have connected it to your machine...

1. Launch Chrome.
2. In the address bar, type `chrome://inspect/#devices`.
3. You should see at least one entry for either `Keyman` or for its `SystemKeyboard` under "Remote Target".
4. Look for an entry with a URL including `keyboard.html` and choose to "inspect" it.

According to https://medium.com/@xabaras/remotely-debugging-android-webview-content-3bff59b14ab1, this may require your Android device be at least Android 4.4.  (That link has a few example images that should be pretty current.)

**Comment by MattGyverLee** [comment]
> ```
> .ios kmw-keyboard-sil_cameroon_qwerty #shift-T_0330 {
> }
> 
> .android kmw-keyboard-sil_cameroon_qwerty #shift-T_0330,
> .android kmw-keyboard-sil_cameroon_qwerty::sil_cameroon_qwerty #shift-T_0330, /* support Keyman 10.0-13.0 bug #3910 */
> .android kmw-keyboard-cloud::sil_cameroon_qwerty #shift-T_0330 /* support Keyman 13.0 bug #3910 */
> {
>   // css
> }
> ```

This was close, except that I needed dots before kmw-keyboard since it is a class, not an element. Maybe that was part of my problem all along. Rookie CSS mistake.

I'm testing the following:
```
.ios .kmw-keyboard-sil_cameroon_qwerty #shift-T_0330 {
 } 
 .android .kmw-keyboard-sil_cameroon_qwerty #shift-T_0330,
 .android .kmw-keyboard-sil_cameroon_qwerty::sil_cameroon_qwerty #shift-T_0330, /* support Keyman 10.0-13.0 bug #3910 */
 .android .kmw-keyboard-cloud::sil_cameroon_qwerty #shift-T_0330 /* support Keyman 13.0 bug #3910 */
 {
   // css
 }

**Comment by MattGyverLee** [comment]
> According to https://medium.com/@xabaras/remotely-debugging-android-webview-content-3bff59b14ab1, this may require your Android device be at least Android 4.4. (That link has a few example images that should be pretty current.)

Thanks, Android default fonts only fully supported the Cameroon Orthography (mainly the capital Ə and Ɔ) starting with 5.1.1, so I'm not supporting anything lower than that anyway.  I can fix the OSK font, but I don't want to get a flood of complaints where output characters don't display in many apps when there's nothing I can do.

**Comment by MattGyverLee** [comment]
Uh oh, it gets more complex. Colon needs to be escaped in CSS selectors. I haven't found a working combination yet (\: or \3a).

**Comment by mcdurdin** [comment]
`\:\:` should work -- I just tested here and it worked. I've updated my comment to address both the missing period and the backslashes.

**Comment by MattGyverLee** [comment]
This will be closed by https://github.com/keymanapp/keyboards/pull/1394 . Thanks all!
